### DISCUSSION TITLE
How can I test type conversion without raising exceptions?

### DISCUSSION THREAD

**Discussion by nat45928** [nlohmann/json#2385] [Q&A]
I have an application that does not have exceptions enabled. How can I test whether a type is convertible without raising an exception? For example:

```
json data = {
         {"boolean", true},
         {"string", "foo},
};

bool b = data.get<bool>("string"); // throws
```

How can I test `get("string")` is convertible to `b`s type (bool in this case)?

I've poked around and realized that I can use the `.is_*()` methods for basic types but objects and arrays are much more complicated. I'm hoping to find a general-ish solution if possible.

Thanks for any suggestions!

**Comment by nlohmann** [comment]
You can check the type via functions like [`is_string`](https://nlohmann.github.io/json/api/basic_json/is_string/) or [`is_boolean `](https://nlohmann.github.io/json/api/basic_json/is_boolean/).

**Reply by nat45928** [reply]
Thanks for the tip!

How does that extrapolate to data structures like objects and arrays? I.e. How can I check if a `json` object is convertible to `std::vector<std::pair<double, double>>`?

I see there are a few functions relating to this like `from_json_array_impl()`. I was hoping there is a way to reuse some of the logic existing in the library instead of just rolling my own logic.

**Reply by nlohmann** [reply]
See #2404.
